### PR TITLE
feat: support GatewayClass, refactor gateway modules

### PR DESCRIPTION
**Pull request by lingsamuel**
Signed-off-by: Ling Samuel <lingsamuelgrace@gmail.com>

### Type of change:

- [x] New feature provided

### CODE REVIEW

**Comment by tao12345666333** [comment]
please fix CI errors.

File: pkg/ingress/namespace/namespace.go [hunk 1]
```diff
@@ -22,21 +22,28 @@ import (
...
-	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
 
 	"github.com/apache/apisix-ingress-controller/pkg/log"
 	"github.com/apache/apisix-ingress-controller/pkg/types"
 )
 
+// FIXME: Controller should be the Core Part,
+// Provider should act as "EventHandler", register there functions to Controller
```
Comment (tao12345666333) [inline pkg/ingress/namespace/namespace.go:33]: Will it be implemented in a follow-up PR?

File: pkg/ingress/namespace/provider.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+func (c *watchingProvider) WatchingNamespaces() []string {
+	var keys []string
+	c.watchingNamespaces.Range(func(key, _ interface{}) bool {
+		keys = append(keys, key.(string))
+		return true
+	})
+	return keys
+}
+
+// isWatchingNamespace accepts a resource key, getting the namespace part
```
Comment (tao12345666333) [inline pkg/ingress/namespace/provider.go:155]:
```suggestion
// IsWatchingNamespace accepts a resource key, getting the namespace part
```

Comment (tao12345666333) [inline pkg/ingress/namespace/provider.go:155]:
```suggestion
// IsWatchingNamespace accepts a resource key, getting the namespace part
```
```suggestion
// isWatchingNamespace accepts a resource key, getting the namespace part
```

**Review by tao12345666333 — APPROVED** [review]

File: pkg/ingress/namespace/namespace.go (hunk 1 above)
Comment (lingsamuel) [reply pkg/ingress/namespace/namespace.go:33]: Yes. The core concept here is that the providers shouldn't interact directly with the controller's field, for example add event handlers directly to the informers.

**Review by gxthrj — APPROVED** [review]

**Pull request merged**